source/language/openpulse.rst

```diff
@@ -0,0 +1,609 @@
+  frame driveframe = newframe(5e9, 0.0); // newframe(float[size] frequency, angle[size] phase)
+
+would construct a frame with frequency `f`  5 GHz and phase :math:`\theta` 0.0. When
+instantiated, the frame time starts at 0. ``frame``s can also be copied using the
+``copyframe`` command
+
+.. code-block:: javascript
+
+  frame driveframe1 = newframe(5e9, 0.0);
+  frame driveframe2 = copyframe(driveframe1);
+  driveframe2.phase = driveframe1.phase + pi/2;
+
+This will generate a ::math:`pi/2` phase incremented copy of ``driveframe1`` with the same
+time as `driveframe1`.
+
+Frame manipulation
+~~~~~~~~~~~~~~~~~~
+
+The (frequency, phase) tuple of a frame can be manipulated throughout program
+by referencing ``.frequency``, and ``.phase``. Operations must be
+appropriate for the respective type, ``float`` for frequency and ``angle`` for
+phase. Again, the exact precision of these calculations is hardware specific.
+
+Here's an example of manipulating the phase to calibrate an ``rz`` gate on a frame called
+``driveframe``:
+
+.. code-block:: javascript
+
+   // Example 1: Shift phase of the "drive" frame by pi/4, eg. an rz gate with angle -pi/4
+   cal {
+     driveframe.phase += pi/4;
+   }
+
+   // Example 2: Define a calibration for the rz gate on all 8 physical qubits
+   cal {
+     frame[8] rz_frames;
+     frame[0] = newframe(...);
+     // and so on
+   }
+
+   defcal rz(angle[20] theta) $q {
+     rz_frames[q].phase -= theta;
+   }
+
+Manipulating frames based on the state of other frames is also permitted:
+
+.. code-block:: javascript
+
+   // Swap phases between two frames
+   const temp = frame1.phase;
+   frame1.phase = frame2.phase;
+   frame2.phase = temp;
+
+Waveforms
+---------
+
+Waveforms are of type ``waveform`` and can either be:
+
+- An array of complex samples which define the points for the waveform envelope
+- An abstract mathematical function representing a waveform. This will later be
+  materialized into a list of complex samples, either by the compiler or the hardware
+  using the parameters provided to the pulse template.
+
+A value of type ``waveform`` is retrieved by explicitly constructing the complex samples
+or by calling one of the built-in waveform template functions. The latter are initialized by
+assigning a ``waveform`` to the result of a (kernel) function call. Note that each of these
+kernel functions takes a type ``length`` as a first argument, since waveforms need to have a
+definite length. Using the hardware dependent ``dt`` unit is recommended for this length,
+since the compiler may need to down-sample a higher precision waveform to physically realize it.
+
+.. code-block:: javascript
+
+   // arbitrary complex samples
+   waveform arb_waveform = [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // sigma is the standard deviation of waveform
+   kernel gaussian(complex[size] amp, length l, length sigma) -> waveform;
```

Comment:
`kernel` here and below should be `extern`.

Reply:
Resolved in e388e06.